### PR TITLE
Add table aws_sagemaker_notebook_instance. closes #200

### LINKED ISSUE #200: Add table aws_sagemaker_notebook_instance

### PR DESCRIPTION
**Pull request by Priyanka585464**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
SETUP: tests/aws_sagemaker_notebook_instance []

PRETEST: tests/aws_sagemaker_notebook_instance

TEST: tests/aws_sagemaker_notebook_instance
Running terraform
data.aws_region.alternate: Refreshing state...
data.aws_region.primary: Refreshing state...
data.aws_caller_identity.current: Refreshing state...
data.aws_partition.current: Refreshing state...
data.null_data_source.resource: Refreshing state...
aws_iam_role.my_role: Creating...
aws_iam_role.my_role: Creation complete after 6s [id=turbottest83303]
aws_sagemaker_notebook_instance.named_test_resource: Creating...
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [10s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [20s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [30s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [40s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [50s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [1m0s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [1m10s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [1m20s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [1m30s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [1m40s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [1m50s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [2m0s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [2m10s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [2m20s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [2m30s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [2m40s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [2m50s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [3m0s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [3m10s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [3m20s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [3m30s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [3m40s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [3m50s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Still creating... [4m0s elapsed]
aws_sagemaker_notebook_instance.named_test_resource: Creation complete after 4m3s [id=turbottest83303]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

Outputs:

account_id = 986325076436
aws_partition = aws
aws_region = us-east-1
resource_aka = arn:aws:sagemaker:us-east-1:986325076436:notebook-instance/turbottest83303
resource_id = turbottest83303
resource_name = turbottest83303

Running SQL query: test-get-query.sql
[
  {
    "account_id": "986325076436",
    "arn": "arn:aws:sagemaker:us-east-1:986325076436:notebook-instance/turbottest83303",
    "name": "turbottest83303",
    "partition": "aws",
    "region": "us-east-1",
    "tags": {
      "name": "turbottest83303"
    }
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "akas": [
      "arn:aws:sagemaker:us-east-1:986325076436:notebook-instance/turbottest83303"
    ],
    "name": "turbottest83303",
    "tags": {
      "name": "turbottest83303"
    }
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "arn": "arn:aws:sagemaker:us-east-1:986325076436:notebook-instance/turbottest83303",
    "name": "turbottest83303",
    "partition": "aws",
    "region": "us-east-1"
  }
]
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "account_id": "986325076436",
    "akas": [
      "arn:aws:sagemaker:us-east-1:986325076436:notebook-instance/turbottest83303"
    ],
    "region": "us-east-1",
    "tags": {
      "name": "turbottest83303"
    },
    "title": "turbottest83303"
  }
]
✔ PASSED

POSTTEST: tests/aws_sagemaker_notebook_instance

TEARDOWN: tests/aws_sagemaker_notebook_instance

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
### Basic info

```sql
select
  name,
  arn,
  creation_time,
  instance_type,
  notebook_instance_status
from
  aws_sagemaker_notebook_instance;

```
```
+------------------------------------+-----------------------------------------------------------------------------------------------+---------------------+--
| name                               | arn                                                                                           | creation_time       | i
+------------------------------------+-----------------------------------------------------------------------------------------------+---------------------+--
| steampipe12                        | arn:aws:sagemaker:us-east-1:986325076436:notebook-instance/steampipe12                        | 2021-04-16 08:17:51 | m
| test-rupzz                         | arn:aws:sagemaker:us-east-1:986325076436:notebook-instance/test-rupzz                         | 2021-04-08 09:22:53 | m
| turbot-test-20200125-create-update | arn:aws:sagemaker:us-east-1:986325076436:notebook-instance/turbot-test-20200125-create-update | 2021-04-16 09:32:26 | m
+------------------------------------+-----------------------------------------------------------------------------------------------+---------------------+--

```
### List notebook instances with unecrypted data on the ML storage volume

```sql
select
  name,
  kms_key_id
from
  aws_sagemaker_notebook_instance
where
  kms_key_id is null;
```
```
+------------------------------------+------------+
| name                               | kms_key_id |
+------------------------------------+------------+
| turbot-test-20200125-create-update | <null>     |
| test-rupzz                         | <null>     |
+------------------------------------+------------+
```

### List notebook instances that are not publicly available

```sql
select
  name,
  direct_internet_access
from
  aws_sagemaker_notebook_instance
where
  direct_internet_access = 'Disabled';
```
```
+------+------------------------+
| name | direct_internet_access |
+------+------------------------+
+------+------------------------+
```

### List notebook instances whose root access for users disabled

```sql
select
  name,
  root_access
from
  aws_sagemaker_notebook_instance
where
  root_access = 'Disabled';
```
```
+------+-------------+
| name | root_access |
+------+-------------+
+------+-------------+
```

### CODE REVIEW

File: aws-test/tests/aws_sagemaker_notebook_instance/test-get-expected.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+		"partition": "{{ output.aws_partition.value }}",
+		"region": "{{ output.aws_region.value }}",
+		"tags": [
+      {
+        "Key": "name",
+        "Value": "{{ resourceName }}"
+      }
+    ]
+	}
+]
```
Comment (Paulami30) [inline aws-test/tests/aws_sagemaker_notebook_instance/test-get-expected.json:15]: fix indentation

File: aws/table_aws_sagemaker_notebook_instance.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+				Transform:   transform.FromField("NotebookInstanceArn"),
+			},
+			{
+				Name:        "creation_time",
+				Description: "A timestamp that shows when the notebook instance was created.",
+				Type:        proto.ColumnType_TIMESTAMP,
+			},
+			{
+				Name:        "default_code_repository",
+				Description: "The date and time that the cluster was created.",
```
Comment (Paulami30) [inline aws/table_aws_sagemaker_notebook_instance.go:48]: is the description correct?

File: aws/table_aws_sagemaker_notebook_instance.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+			},
+			{
+				Name:        "url",
+				Description: "The URL that you use to connect to the Jupyter notebook that is running in your notebook instance.",
+				Type:        proto.ColumnType_STRING,
+				Hydrate:     getAwsSageMakerNotebookInstance,
+			},
+			{
+				Name:        "volume_size_in_gb",
+				Description: "Describes a group of DeferredMaintenanceWindow objects.",
```
Comment (Paulami30) [inline aws/table_aws_sagemaker_notebook_instance.go:121]: check the description again

File: aws/table_aws_sagemaker_notebook_instance.go [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     getAwsSageMakerNotebookInstance,
+			},
+			{
+				Name:        "tags_src",
+				Description: "The list of tags for the notebook instance.",
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     listAwsSageMakerNotebookInstanceTags,
+				Transform:   transform.FromField("Tags"),
+			},
```
Comment (Paulami30) [inline aws/table_aws_sagemaker_notebook_instance.go:148]:
```suggestion
			},
			
```

File: aws/table_aws_sagemaker_notebook_instance.go [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("NotebookInstanceName"),
+			},
+			{
+				Name:        "akas",
+				Description: resourceInterfaceDescription("akas"),
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     getAwsSageMakerNotebookInstanceAka,
+				Transform:   transform.FromValue(),
+			},
```
Comment (Paulami30) [inline aws/table_aws_sagemaker_notebook_instance.go:168]: arrange the columns as title, tags, akas

File: aws/table_aws_sagemaker_notebook_instance.go [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+	// Get call
+	op, err := svc.ListTags(params)
+	if err != nil {
+		logger.Debug("listAwsSageMakerNotebookInstanceTags", "ERROR", err)
+		return nil, err
+	}
+
+	return op, nil
+}
+
```
Comment (Paulami30) [inline aws/table_aws_sagemaker_notebook_instance.go:273]: you are already getting the arn of the notebook instance.  why is this function required?

File: docs/tables/aws_sagemaker_notebook_instance.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  name,
+  kms_key_id
+from
+  aws_sagemaker_notebook_instance
+where
+  kms_key_id is null;
+```
+
+
+### List notebook instances that are not publicly available
```
Comment (Paulami30) [inline docs/tables/aws_sagemaker_notebook_instance.md:35]:
```suggestion
### List notebook instances that are publicly available
```

File: docs/tables/aws_sagemaker_notebook_instance.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  name,
+  direct_internet_access
+from
+  aws_sagemaker_notebook_instance
+where
+  direct_internet_access = 'Disabled';
+```
+
+
+### List notebook instances whose root access for users disabled
```
Comment (Paulami30) [inline docs/tables/aws_sagemaker_notebook_instance.md:48]:
```suggestion
### List notebook instances which provides root access to the users
```

Comment (Paulami30) [reply docs/tables/aws_sagemaker_notebook_instance.md:48]: Also make changes in the query accordingly

**Review by Paulami30 — CHANGES_REQUESTED** [review]
Please check for the comments

File: aws-test/tests/aws_sagemaker_notebook_instance/test-get-expected.json [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+		"partition": "{{ output.aws_partition.value }}",
+		"region": "{{ output.aws_region.value }}",
+		"tags": [
+			{
+				"Key": "name",
+				"Value": "{{ resourceName }}"
+			}
+		]
+	}
+]
```
Comment (Paulami30) [inline aws-test/tests/aws_sagemaker_notebook_instance/test-get-expected.json:15]: can you check the indentation for this again?

File: aws-test/tests/aws_sagemaker_notebook_instance/test-turbot-expected.json [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+		"region": "{{ output.aws_region.value }}",
+		"tags": [
+      {
+        "Key": "name",
+        "Value": "{{ resourceName }}"
+      }
+    ],
+		"title": "{{ resourceName }}"
+	}
+]
```
Comment (Paulami30) [inline aws-test/tests/aws_sagemaker_notebook_instance/test-turbot-expected.json:14]: fix indentation

File: aws/table_aws_sagemaker_notebook_instance.go [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+				Name:        "tags",
+				Description: resourceInterfaceDescription("tags"),
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     listAwsSageMakerNotebookInstanceTags,
+			},
+			{
+				Name:        "akas",
+				Description: resourceInterfaceDescription("akas"),
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("NotebookInstanceArn"),
```
Comment (Paulami30) [inline aws/table_aws_sagemaker_notebook_instance.go:167]: use `arnToAkas` function here

File: aws/table_aws_sagemaker_notebook_instance.go [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+// 	c, err := getCommonColumns(ctx, d, h)
+// 	if err != nil {
+// 		return nil, err
+// 	}
+// 	commonColumnData := c.(*awsCommonColumnData)
+// 	aka := []string{"arn:" + commonColumnData.Partition + ":sagemaker:" + commonColumnData.Region + ":" + commonColumnData.AccountId + ":notebook-instance/" + NotebookInstanceName}
+
+// 	return aka, nil
+// }
+
```
Comment (Paulami30) [inline aws/table_aws_sagemaker_notebook_instance.go:286]: remove the commented part

File: aws/table_aws_sagemaker_notebook_instance.go [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+				Description: resourceInterfaceDescription("title"),
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("NotebookInstanceName"),
+			},
+			{
+				Name:        "tags",
+				Description: resourceInterfaceDescription("tags"),
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     listAwsSageMakerNotebookInstanceTags,
+			},
```
Comment (Paulami30) [inline aws/table_aws_sagemaker_notebook_instance.go:162]: this is tags in the original format. why did not you convert it to turbot tags?

File: docs/tables/aws_sagemaker_notebook_instance.md [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+```sql
+select
+  name,
+  kms_key_id
+from
+  aws_sagemaker_notebook_instance
+where
+  kms_key_id is null;
+```
+
```
Comment (Paulami30) [inline docs/tables/aws_sagemaker_notebook_instance.md:46]:
```suggestion


```

**Review by Paulami30 — CHANGES_REQUESTED** [review]
Check the comments, run the integration test and update the latest log

File: aws-test/tests/aws_sagemaker_notebook_instance/test-turbot-expected.json [hunk 15]
```diff
@@ -0,0 +1,11 @@
+[
+	{
+		"account_id": "{{ output.account_id.value }}",
+		"akas": ["{{ output.resource_aka.value }}"],
+		"region": "{{ output.aws_region.value }}",
+		"tags": {
+      "name": "{{ resourceName }}"
```
Comment (Paulami30) [inline aws-test/tests/aws_sagemaker_notebook_instance/test-turbot-expected.json:7]: wrong indentation

File: aws/table_aws_sagemaker_notebook_instance.go [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+		return *item.(*sagemaker.DescribeNotebookInstanceOutput).NotebookInstanceName
+	}
+	return ""
+}
+
+//// TRANSFORM FUNCTIONS
+
+func getAwsSageMakerNotebookInstanceTurbotTags(ctx context.Context, d *transform.TransformData) (interface{},
+	error) {
+	instanceTags := d.HydrateItem.(*sagemaker.ListTagsOutput)
```
Comment (Paulami30) [inline aws/table_aws_sagemaker_notebook_instance.go:289]:
```suggestion
	data := d.HydrateItem.(*sagemaker.ListTagsOutput)
```

File: aws/table_aws_sagemaker_notebook_instance.go [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+
+	if instanceTags.Tags == nil {
```
Comment (Paulami30) [inline aws/table_aws_sagemaker_notebook_instance.go:291]:
```suggestion
	if data.Tags == nil {
```

File: aws/table_aws_sagemaker_notebook_instance.go [hunk 18]
```diff
@@ -0,0 +1,303 @@
...
+	switch item.(type) {
+	case *sagemaker.NotebookInstanceSummary:
+		return *item.(*sagemaker.NotebookInstanceSummary).NotebookInstanceName
+	case *sagemaker.DescribeNotebookInstanceOutput:
+		return *item.(*sagemaker.DescribeNotebookInstanceOutput).NotebookInstanceName
+	}
+	return ""
+}
+
+//// TRANSFORM FUNCTIONS
```
Comment (Paulami30) [inline aws/table_aws_sagemaker_notebook_instance.go:285]:
```suggestion
//// TRANSFORM FUNCTION
```

File: aws/table_aws_sagemaker_notebook_instance.go [hunk 19]
```diff
@@ -0,0 +1,303 @@
...
+func instanceName(item interface{}) string {
+	switch item.(type) {
+	case *sagemaker.NotebookInstanceSummary:
+		return *item.(*sagemaker.NotebookInstanceSummary).NotebookInstanceName
+	case *sagemaker.DescribeNotebookInstanceOutput:
+		return *item.(*sagemaker.DescribeNotebookInstanceOutput).NotebookInstanceName
+	}
+	return ""
+}
+
```
Comment (Paulami30) [inline aws/table_aws_sagemaker_notebook_instance.go:284]: move the above function after the transform function

**Review by Paulami30 — APPROVED** [review]

**Comment by Paulami30** [comment]
LGTM

**Review by rajmohanty17 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**